megamek/src/megamek/client/bot/princess/Princess.java

```diff
@@ -283,6 +283,27 @@ public PathRankerState getPathRankerState() {
     Precognition getPrecognition() {
         return precognition;
     }
+    
+    public int getMaxWeaponRange(Entity entity) {
+        return getMaxWeaponRange(entity, false);
+    }
+    
+    /**
+     * Retrieves maximum weapon range for the given entity.
+     * Cached version of entity.getMaxWeaponRange() 
+     * @param entity Entity we're checking
+     * @param airborneTarget Whether the potential target is in the air, only relevant for
+     *          aircraft shooting at other aircraft on ground maps.
+     * @return
+     */
+    public int getMaxWeaponRange(Entity entity, boolean airborneTarget) {
+        if(!getFireControlState().getWeaponRanges(airborneTarget).containsKey(entity.getId())) {
```

Comment:
You could consider something like:
```java
return getFireControlState().getWeaponRanges(airborneTarget)
    .computeIfAbsent(entity.getId(), _ -> entity.getMaxWeaponRange(airborneTarget));
```

Reply:
Did not know that was a thing. I'll take a look at it.

Reply:
BTW: Java complains about underscore as an identifier, so I guess make that `id` instead.